### PR TITLE
Implement submission locks

### LINKED ISSUE #595: Submission Lock
A `BooleanField` on each submission, that would prevent it from being rejudged. Should probably be applied to contest submissions post-contest (automatically?)

Would probably require two new perms as well:
- `RejudgeLockedSubmission` name is self-explanatory
- `UnlockSubmission` allows user to unlock a sub

### PR DESCRIPTION
**Pull request by WallE256**
Closes #595

### CODE REVIEW

File: judge/models/contest.py [hunk 1]
```diff
@@ -126,6 +126,8 @@ class Contest(models.Model):
                                             help_text='A custom Lua function to generate problem labels. Requires a '
                                                       'single function with an integer parameter, the zero-indexed '
                                                       'contest problem index, and returns a string, the label.')
+    is_locked = models.BooleanField(verbose_name=_('contest lock'), default=False,
+                                    help_text='If locked, the submissions from this contest will be locked')
```
Comment (Carson-Tang) [inline judge/models/contest.py:130]: Period

File: judge/admin/contest.py [hunk 2]
```diff
@@ -211,13 +216,48 @@ def make_hidden(self, request, queryset):
...
             url(r'^(\d+)/rate/$', self.rate_view, name='judge_contest_rate'),
             url(r'^(\d+)/judge/(\d+)/$', self.rejudge_view, name='judge_contest_rejudge'),
+            url(r'^(\d+)/lock/$', self.lock_view, name='judge_contest_lock'),
         ] + super(ContestAdmin, self).get_urls()
 
+    def lock_view(self, request, id):
+        if not request.user.has_perm('judge.contest_lock'):
+            raise PermissionDenied()
+        contest = get_object_or_404(Contest, id=id)
+        self.set_is_locked(contest, not contest.is_locked)
```
Comment (quantum5) [inline judge/admin/contest.py:258]: Not sure how I feel about the lack of idempotence with this toggle, and the fact it's a GET.

File: judge/models/contest.py [hunk 3]
```diff
@@ -126,6 +126,8 @@ class Contest(models.Model):
                                             help_text='A custom Lua function to generate problem labels. Requires a '
                                                       'single function with an integer parameter, the zero-indexed '
                                                       'contest problem index, and returns a string, the label.')
+    is_locked = models.BooleanField(verbose_name=_('contest lock'), default=False,
+                                    help_text='If locked, the submissions from this contest will be locked.')
```
Comment (TheAvidDev) [inline judge/models/contest.py:130]: Either this should explain what:
> submissions from this contest will be locked

means, or `help_text` should be included in the `is_locked` field of a Submission.

File: judge/admin/contest.py [hunk 4]
```diff
@@ -211,13 +216,48 @@ def make_hidden(self, request, queryset):
...
+        self.message_user(request, ungettext('%d contest successfully unlocked.',
+                                             '%d contests successfully unlocked.',
+                                             count) % count)
+    set_unlocked.short_description = _('Unlock contest submissions')
+
+    def set_is_locked(self, contest, is_locked):
+        with transaction.atomic():
+            contest.is_locked = is_locked
+            contest.save()
+            for row in ContestSubmission.objects.filter(participation__contest=contest):
```
Comment (quantum5) [inline judge/admin/contest.py:241]: Why do we need this loop? `ContestSubmission.objects.filter(participation__contest=contest).update(submission__is_locked=True)` works just as well.

File: judge/migrations/0106_submission_lock.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+def updatecontestsubmissions(apps, schema_editor):
+    Contest = apps.get_model('judge', 'Contest')
+    for row in Contest.objects.all():
+        if row.end_time < timezone.now():
+            row.is_locked = True
+            row.save()
+
+    ContestSubmissions = apps.get_model('judge', 'ContestSubmission')
+    for row in ContestSubmissions.objects.all():
```
Comment (quantum5) [inline judge/migrations/0106_submission_lock.py:15]: This can be done in one query.

File: judge/admin/contest.py [hunk 6]
```diff
@@ -211,13 +216,45 @@ def make_hidden(self, request, queryset):
...
+        self.message_user(request, ungettext('%d contest successfully unlocked.',
+                                             '%d contests successfully unlocked.',
+                                             count) % count)
+    set_unlocked.short_description = _('Unlock contest submissions')
+
+    def set_is_locked(self, contest, is_locked):
+        with transaction.atomic():
+            contest.is_locked = is_locked
+            contest.save()
+            ContestSubmission.objects.filter(participation__contest=contest).update(submission__is_locked=is_locked)
```
Comment (quantum5) [inline judge/admin/contest.py:241]: Does this work? Didn't you use something else in the migration?

File: templates/admin/judge/contest/change_form.html [hunk 7]
```diff
@@ -20,4 +20,11 @@
             <span class="text">{% trans "Rate" %}</span>
         </a>
     {% endif %}
+    {% if original and perms.judge.contest_lock %}
+        <a style="display: none" title="{% trans "Toggle Lock" %}" href="{% url 'admin:judge_contest_lock' original.pk %}"
+           class="button rerate-link">
```
Comment (Ninjaclasher) [inline templates/admin/judge/contest/change_form.html:25]: `rerate-link`?

File: judge/migrations/0106_submission_lock.py [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+def updatecontestsubmissions(apps, schema_editor):
+    Contest = apps.get_model('judge', 'Contest')
+    for row in Contest.objects.all():
+        if row.end_time < timezone.now():
+            row.is_locked = True
+            row.save()
+
+    Submission = apps.get_model('judge', 'Submission')
+    Submission.objects.filter(contest__participation__contest__is_locked=True).update(is_locked=True)
```
Comment (Ninjaclasher) [inline judge/migrations/0106_submission_lock.py:15]:
```suggestion
    Submission.objects.filter(contest_object__is_locked=True).update(is_locked=True)
```

File: judge/migrations/0106_submission_lock.py [hunk 9]
```diff
@@ -0,0 +1,44 @@
+# Generated by Django 2.2.12 on 2020-05-13 20:58
+
+from django.db import migrations, models
+from django.utils import timezone
+
+
+def updatecontestsubmissions(apps, schema_editor):
+    Contest = apps.get_model('judge', 'Contest')
+    for row in Contest.objects.all():
```
Comment (Ninjaclasher) [inline judge/migrations/0106_submission_lock.py:9]: Similar thing as with the contest admin action, is there a reason a `Contest.objects.filter(...).update(...)` can't be used here?

File: judge/admin/contest.py (hunk 6 above)
Comment (Ninjaclasher) [inline judge/admin/contest.py:241]: Would it be more efficient to use a `Submission.objects.filter(contest_object=contest).update(is_locked=True)` here instead?

File: judge/migrations/0106_submission_lock.py (hunk 9 above)
Comment (WallE256) [reply judge/migrations/0106_submission_lock.py:9]: Only base field are available in migrations, so the `ended` property has to be manually computed.

Comment (Ninjaclasher) [reply judge/migrations/0106_submission_lock.py:9]: `ended` doesn't seem to be used anywhere here. Only `end_time` and `is_locked` are used, which are both base fields (unless I'm misunderstanding something).

File: judge/admin/contest.py (hunk 2 above)
Comment (quantum5) [reply judge/admin/contest.py:255]: This should at least be a post request, and might make more sense have separate lock and unlock operations.

File: judge/models/contest.py [hunk 10]
```diff
@@ -342,6 +344,7 @@ class Meta:
             ('create_private_contest', _('Create private contests')),
             ('change_contest_visibility', _('Change contest visibility')),
             ('contest_problem_label', _('Edit contest problem label script')),
+            ('lock_contest', 'Change lock status of contest'),
```
Comment (Ninjaclasher) [inline judge/models/contest.py:347]:
```suggestion
            ('lock_contest', _('Change lock status of contest')),
```

File: judge/models/contest.py (hunk 3 above)
Comment (WallE256) [reply judge/models/contest.py:130]: Maybe something like `Prevent submissions from this contest from being rejudged`?

Comment (TheAvidDev) [reply judge/models/contest.py:130]: Sounds good.

File: judge/models/submission.py [hunk 11]
```diff
@@ -114,6 +115,8 @@ def long_status(self):
         return Submission.USER_DISPLAY_CODES.get(self.short_status, '')
 
     def judge(self, *args, **kwargs):
+        if self.is_locked:
+            return
         judge_submission(self, *args, **kwargs)
```
Comment (Xyene) [inline judge/models/submission.py:120]:
```suggestion
        if not self.is_locked:
            judge_submission(self, *args, **kwargs)
```

File: templates/admin/judge/contest/change_form.html [hunk 12]
```diff
@@ -20,4 +21,18 @@
             <span class="text">{% trans "Rate" %}</span>
         </a>
     {% endif %}
+    {% if original and not original.is_locked and perms.judge.contest_lock %}
+        <a style="display: none" title="{% trans "Lock Contest" %}" href="{% url 'admin:judge_contest_lock' original.pk %}"
+           class="button lock-link">
+            <i class="fa fa-lg fa-lock"></i>
+            <span class="text">{% trans "Lock Contest" %}</span>
```
Comment (Xyene) [inline templates/admin/judge/contest/change_form.html:28]:
```suggestion
            <span class="text">{% trans "Lock submissions" %}</span>
```

File: templates/admin/judge/contest/change_form.html [hunk 13]
```diff
@@ -20,4 +21,18 @@
...
+        </a>
+    {% endif %}
+    {% if original and original.is_locked and perms.judge.contest_lock %}
+        <a style="display: none" title="{% trans "Unlock Contest" %}" href="{% url 'admin:judge_contest_unlock' original.pk %}"
+           class="button lock-link">
+            <i class="fa fa-lg fa-unlock"></i>
+            <span class="text">{% trans "Unlock Contest" %}</span>
```
Comment (Xyene) [inline templates/admin/judge/contest/change_form.html:35]:
```suggestion
            <span class="text">{% trans "Unlock submissions" %}</span>
```

File: judge/admin/contest.py [hunk 14]
```diff
@@ -211,13 +216,53 @@ def make_hidden(self, request, queryset):
...
+        self.message_user(request, ungettext('%d contest successfully unlocked.',
+                                             '%d contests successfully unlocked.',
+                                             count) % count)
+    set_unlocked.short_description = _('Unlock contest submissions')
+
+    def set_is_locked(self, contest, is_locked):
+        with transaction.atomic():
+            contest.is_locked = is_locked
+            contest.save()
+            Submission.objects.filter(contest_object=contest).update(is_locked=is_locked)
```
Comment (kiritofeng) [inline judge/admin/contest.py:241]: Should we lock all submissions, or only non-virtual submissions?

**Pull request merged**